### PR TITLE
[tnx] add vllm rolling batch implementation

### PR DESCRIPTION
**Pull request by tosterberg**
## Description ##

This PR adds vLLM support in Neuron in two capacities.
- First as a passthrough backend where vllm is used for model loading, and rolling batch scheduling
- Second as a rolling batch scheduler only where model loading is handled by current solutions

Testing for Neuron has been moved to its own folder and is now unit tested as part of the integration tests, with the long term goal of providing a location for others to test our neuron integration without the container or model server. This requires us to be able to run unit tests in context, and is supported by this change.

- If this change is a backward incompatible change, why must this change be made?
- Interesting edge cases to note here

### CODE REVIEW

File: engines/python/setup/djl_python/neuron_utils/model_loader.py [hunk 1]
```diff
@@ -143,15 +144,26 @@ def set_neuronx_class(self):
...
+        Set configuration for continuous batching, currently all vllm implementations are continuous batching
+        and batch size greater than 1 for tnx and lmi-dist support rolling batch.
+
+        :return: bool indicating if continuous batching can be used
+        """
+        use_continuous_batching = (self.config.rolling_batch != "disable"
+                                   and self.config.rolling_batch_strategy
+                                   == TnXGenerationStrategy.continuous_batching
+                                   and self.config.max_rolling_batch_size
+                                   > 1) or self.config.rolling_batch == "vllm"
```
Comment (siddvenk) [inline engines/python/setup/djl_python/neuron_utils/model_loader.py:158]: does lmi-dist need to be handled here?

**Review by sindhuvahinis — APPROVED** [review]

**Comment by sindhuvahinis** [comment]
No major comments now. Just seems like a lot of repetitive code between lmi-dist, vllm and now with neuron-vllm. I am doing refactor with vllm_utils to support RequestOutput, that should address this a lit bit. 

As we discussed offline, Approving it for now. You can merge it after addressing Sid's comment.

File: engines/python/setup/djl_python/neuron_utils/model_loader.py (hunk 1 above)
Comment (tosterberg) [reply engines/python/setup/djl_python/neuron_utils/model_loader.py:158]: Currently no, as of right now lmi-dist does not have this model compilation limitation, so this is base vLLM support only that requires continuous batching for batch size 1 (reason we don't use this is that the performance is worse compared to the alternative scheduler for batch size 1)

File: engines/python/setup/djl_python/rolling_batch/neuron_vllm_rolling_batch.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+            preloaded_model=preloaded_model,
+            tensor_parallel_size=self.vllm_configs.tensor_parallel_degree,
+            dtype=DTYPE_MAPPER[self.vllm_configs.dtype],
+            seed=0,
+            max_model_len=self.vllm_configs.n_positions,
+            max_num_seqs=self.vllm_configs.max_rolling_batch_size,
+            block_size=self.vllm_configs.n_positions,
+            trust_remote_code=self.vllm_configs.trust_remote_code,
+            revision=self.vllm_configs.revision,
+        )
```
[the remark this answers is not in the document]
Comment (tosterberg) [reply engines/python/setup/djl_python/rolling_batch/neuron_vllm_rolling_batch.py:59]: The only other place where we have deviation is the init parameters. Right now the NeuronVLLMRB class init also can take a preloaded model to bypass vllm model loading, which is slow and does not support advanced neuron compilation configs for better performance. I do think we can merge them but it would also require a change to the vllm rb properties in order to allow for a pretrained model as part of the properties, and neuron specific EngineArgs.

I will take a stab at a refactor and see if there is a way to combine.

[the remark this answers is not in the document]
Comment (siddvenk) [reply engines/python/setup/djl_python/rolling_batch/neuron_vllm_rolling_batch.py:59]: I see, I think it would be good to explore this for consolidating the two handlers. But that may take some time, and based on Sindhu's comments seems like this refactor effort is already ongoing and might lead to some other changes. I'm fine with this for now as well, given we probably want it in for the release

**Review by siddvenk — APPROVED** [review]

File: engines/python/setup/djl_python/rolling_batch/neuron_vllm_rolling_batch.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tosterberg) [reply engines/python/setup/djl_python/rolling_batch/neuron_vllm_rolling_batch.py]: Updated to combine them, testing now and if everything is good I will squash and merge.
